app/controllers/declaration/AuthorisationProcedureCodeChoiceController.scala

```diff
@@ -40,29 +43,46 @@ class AuthorisationProcedureCodeChoiceController @Inject()(
 )(implicit ec: ExecutionContext)
     extends FrontendController(mcc) with I18nSupport with ModelCacheable with SubmissionErrors {
 
-  private val validTypes =
-    Seq(DeclarationType.STANDARD, DeclarationType.SUPPLEMENTARY, DeclarationType.SIMPLIFIED, DeclarationType.CLEARANCE)
-
-  def displayPage(mode: Mode): Action[AnyContent] = (authenticate andThen journeyType(validTypes)) { implicit request =>
-    val form = AuthorisationProcedureCodeChoice.form().withSubmissionErrors()
-    request.cacheModel.parties.authorisationProcedureCodeChoice match {
-      case Some(data) => Ok(authorisationProcedureCodeChoice(form.fill(data), mode))
-      case _          => Ok(authorisationProcedureCodeChoice(form, mode))
+  def displayPage(mode: Mode): Action[AnyContent] = (authenticate andThen journeyType) { implicit request =>
+    request.declarationType match {
+      case CLEARANCE if request.cacheModel.isNotEntryIntoDeclarantsRecords =>
+        navigator.continueTo(mode, controllers.declaration.routes.DeclarationHolderRequiredController.displayPage)
+      case OCCASIONAL =>
+        navigator.continueTo(mode, controllers.declaration.routes.DeclarationHolderRequiredController.displayPage)
+      case _ =>
+        val form = AuthorisationProcedureCodeChoice.form().withSubmissionErrors()
+        request.cacheModel.parties.authorisationProcedureCodeChoice match {
+          case Some(data) => Ok(authorisationProcedureCodeChoice(form.fill(data), mode))
+          case _          => Ok(authorisationProcedureCodeChoice(form, mode))
+        }
     }
   }
 
+  private val validTypes = Seq(STANDARD, SUPPLEMENTARY, SIMPLIFIED, CLEARANCE)
+
   def submitForm(mode: Mode): Action[AnyContent] = (authenticate andThen journeyType(validTypes)).async { implicit request =>
     AuthorisationProcedureCodeChoice
       .form()
       .bindFromRequest()
       .fold(
         errors => Future.successful(BadRequest(authorisationProcedureCodeChoice(errors, mode))),
-        updateCache(_).map(_ => navigator.continueTo(mode, routes.DeclarationHolderSummaryController.displayPage))
+        updateCache(_).map(exportsDeclarationUpdated => navigator.continueTo(mode, nextPage(exportsDeclarationUpdated)))
       )
   }
 
   private def updateCache(
     choice: AuthorisationProcedureCodeChoice
   )(implicit request: JourneyRequest[AnyContent]): Future[Option[ExportsDeclaration]] =
     updateExportsDeclarationSyncDirect(_.updateAuthorisationProcedureCodeChoice(choice))
+
+  private def nextPage(exportsDeclarationOpt: Option[ExportsDeclaration]): Mode => Call =
+    (for {
+      exportsDeclaration <- exportsDeclarationOpt
+      authProcedureCodeChoice <- exportsDeclaration.parties.authorisationProcedureCodeChoice.map(_.code)
+
+    } yield (exportsDeclaration.`type`, exportsDeclaration.additionalDeclarationType, authProcedureCodeChoice)) match {
```

Comment:
Nit: I would use `def someFun(exportsDeclaration, authProcedureCodeChoice)` for the `yield` block.

Reply:
Tried that, but the `yield` block ends right before `match` statement. So the pattern matching actually works with the output of this for-comprehension.
Therefore, if I were to extract it into a function, then I would still end up with `Option` type as a result and would need to unwrap it providing default value. Which makes pattern matching redundant.